ray_data_eval/simulator/environment.py

```diff
@@ -309,7 +314,17 @@ def tick(self):
             self.scheduling_policy.tick(self)
         logging.debug(f"[{self}] Tick")
         self._current_tick += 1
-        for executor in self._get_executors_sorted():
+        executors_sorted = self._get_executors_sorted()
+        executors_with_output = [e for e in executors_sorted if e.about_to_finish_and_has_output()]
```

Comment:
Maybe better to bake this into `_get_executors_sorted()`. That function was intended to solve this problem in the first place. The current sorting order is:

1. Prefer executors with lower remaining ticks of the running task
2. If there is a tie, prefer the task with lower net output size (output_size - input_size) so that it will free up memory
3. If there is no running task, it is put to last

You're saying you prefer executors with running tasks that are either 1) have >1 remaining ticks (not finishing in the next tick), or 2) has no output.

I think this might be merged into the current executor sorting order by flipping (1) and (2) in the current sorting order, i.e. by net output size first, then by remaining ticks.

I think the case we discussed was because those `"!"` tasks already have negative remaining_ticks, and will always be preferred over the running tasks. We could also special-case this to make the negative remaining_ticks tasks run last.

Reply:
Updated. please take a look.
I think we should do (2) first to decrease the buffer usage.